File: libs/ironcfg-c/tools/crc_diagnostic.c
/*
 * crc_diagnostic.c
 * Diagnostic tool to print CRC values from ICFX files
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <ironcfg/ironcfg_common.h>

typedef struct {
    uint32_t magic;
    uint8_t flags;
    uint8_t reserved;
    uint16_t header_size;
    uint32_t total_size;
    uint32_t dict_offset;
    uint32_t vsp_offset;
    uint32_t index_offset;
    uint32_t payload_offset;
    uint32_t crc_offset;
    uint32_t payload_size;
    uint32_t dict_size;
    uint32_t vsp_size;
} icfx_header_t;

static uint32_t read_u32_le(const uint8_t* buf, size_t offset) {
    return ((uint32_t)buf[offset]) |
           (((uint32_t)buf[offset + 1]) << 8) |
           (((uint32_t)buf[offset + 2]) << 16) |
           (((uint32_t)buf[offset + 3]) << 24);
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        fprintf(stderr, "usage: crc_diagnostic <file.icfx>\n");
        return 1;
    }

    const char* filePath = argv[1];
    FILE* f = fopen(filePath, "rb");
    if (!f) {
        perror("fopen");
        return 1;
    }

    fseek(f, 0, SEEK_END);
    size_t size = ftell(f);
    fseek(f, 0, SEEK_SET);

    uint8_t* data = (uint8_t*)malloc(size);
    if (!data) {
        fprintf(stderr, "malloc failed\n");
        fclose(f);
        return 1;
    }

    if (fread(data, 1, size, f) != size) {
        fprintf(stderr, "fread failed\n");
        free(data);
        fclose(f);
        return 1;
    }
    fclose(f);

    printf("File: %s\n", filePath);
    printf("Size: %zu bytes\n\n", size);

    if (size < 48) {
        fprintf(stderr, "ERROR: File too small for ICFX header\n");
        free(data);
        return 1;
    }

    /* Parse header */
    printf("Magic: %.4s\n", (const char*)data);
    printf("Flags: 0x%02X\n", data[4]);
    printf("  Bit 0 (LE): %d\n", (data[4] & 0x01) ? 1 : 0);
    printf("  Bit 1 (VSP): %d\n", (data[4] & 0x02) ? 1 : 0);
    printf("  Bit 2 (CRC): %d\n", (data[4] & 0x04) ? 1 : 0);
    printf("  Bit 3 (Index): %d\n", (data[4] & 0x08) ? 1 : 0);

    uint32_t crc_offset = read_u32_le(data, 28);
    printf("\nCRC Offset: %u\n", crc_offset);

    if (crc_offset > 0 && crc_offset + 4 <= size) {
        uint32_t stored_crc = read_u32_le(data, crc_offset);

        /* Compute CRC over [0 .. crc_offset) */
        uint32_t computed_crc = icfg_crc32(data, crc_offset);

        printf("CRC Information:\n");
        printf("  Stored CRC: 0x%08X\n", stored_crc);
        printf("  Computed CRC: 0x%08X\n", computed_crc);
        printf("  Match: %s\n", (stored_crc == computed_crc) ? "YES ✓" : "NO ✗");
    } else {
        printf("No CRC present\n");
    }

    free(data);
    return 0;
}

File: tools/ironconfigtool/CrcDiagnostic.cs
using System;
using System.IO;
using IronConfig;

namespace IronConfigTool;

/// <summary>
/// Diagnostic tool to print CRC information from ICFX files
/// </summary>
public static class CrcDiagnostic
{
    public static void PrintCrcInfo(string filePath)
    {
        byte[] data = File.ReadAllBytes(filePath);
        Console.WriteLine($"File: {filePath}");
        Console.WriteLine($"Size: {data.Length} bytes\n");

        // Parse header
        if (data.Length < 48)
        {
            Console.WriteLine("ERROR: File too small for ICFX header");
            return;
        }

        // Check magic
        string magic = System.Text.Encoding.ASCII.GetString(data, 0, 4);
        Console.WriteLine($"Magic: {magic}");

        byte flags = data[4];
        Console.WriteLine($"Flags: 0x{flags:X2}");
        Console.WriteLine($"  Bit 0 (Little Endian): {(flags & 0x01)}");
        Console.WriteLine($"  Bit 1 (VSP): {(flags & 0x02) >> 1}");
        Console.WriteLine($"  Bit 2 (CRC): {(flags & 0x04) >> 2}");
        Console.WriteLine($"  Bit 3 (Index): {(flags & 0x08) >> 3}\n");

        // Read offsets
        uint crcOffset = ReadUInt32LE(data, 28);
        uint payloadSize = ReadUInt32LE(data, 32);
        uint dictSize = ReadUInt32LE(data, 36);
        uint vspSize = ReadUInt32LE(data, 40);

        Console.WriteLine($"CRC Offset: {crcOffset}");
        Console.WriteLine($"Payload Size: {payloadSize}");
        Console.WriteLine($"Dictionary Size: {dictSize}");
        Console.WriteLine($"VSP Size: {vspSize}\n");

        // If CRC present
        if (crcOffset > 0 && crcOffset + 4 <= data.Length)
        {
            uint storedCrc = ReadUInt32LE(data, crcOffset);
            Console.WriteLine($"CRC Information:");
            Console.WriteLine($"  Stored CRC: 0x{storedCrc:X8}");
            Console.WriteLine($"  CRC covers bytes [0 .. {crcOffset})");

            // Compute CRC over [0 .. crcOffset)
            byte[] dataToHash = new byte[crcOffset];
            Array.Copy(data, 0, dataToHash, 0, (int)crcOffset);
            uint computedCrc = Crc32Ieee.Compute(dataToHash);

            Console.WriteLine($"  Computed CRC: 0x{computedCrc:X8}");
            Console.WriteLine($"  Match: {(storedCrc == computedCrc ? "YES ✓" : "NO ✗")}");
        }
        else
        {
            Console.WriteLine("No CRC present (or invalid CRC offset)");
        }
    }

    private static uint ReadUInt32LE(byte[] data, uint offset)
    {
        if (offset + 4 > data.Length)
            throw new ArgumentException("Offset out of bounds");

        return (uint)(
            data[(int)offset] |
            (data[(int)(offset + 1)] << 8) |
            (data[(int)(offset + 2)] << 16) |
            (data[(int)(offset + 3)] << 24)
        );
    }
}
